velox/exec/Driver.cpp

```diff
@@ -1090,6 +981,72 @@ folly::dynamic Driver::toJson() const {
   return obj;
 }
 
+template <typename Func>
+auto Driver::withDeltaCpuWallTimer(
+    Operator* op,
+    TimingMemberPtr timingMember,
+    Func&& timedFunction) -> decltype(timedFunction()) {
+  // If 'trackOperatorCpuUsage_' is true, create and initialize the timer object
+  // to track cpu and wall time of the timedFunction.
+  if (!trackOperatorCpuUsage_) {
+    return timedFunction();
+  }
+
+  // The delta CpuWallTiming object would be recorded to the corresponding
+  // timingMember upon destruction of the timer when withDeltaCpuWallTimer
+  // ends. The timer is created on the stack to avoid heap allocation
+  auto f = [op, timingMember, this](const CpuWallTiming& elapsedTime) {
+    auto elapsedSelfTime = processLazyTiming(*op, elapsedTime);
+    op->stats().withWLock([&](auto& lockedStats) {
+      (lockedStats.*timingMember).add(elapsedSelfTime);
+    });
+  };
+  DeltaCpuWallTimer<decltype(f)> timer(std::move(f));
+
+  return timedFunction();
+}
+
+void Driver::validateOperatorResult(
+    const RowVectorPtr& result,
+    const Operator& op) {
+  VELOX_CHECK(
+      result->size() > 0,
+      "Operator::getOutput() must return nullptr or a non-empty vector: {}",
+      op.operatorType());
+
+  if (ctx_->queryConfig().validateOutputFromOperators()) {
+    try {
+      result->validate({});
+    } catch (const std::exception& e) {
+      VELOX_FAIL(
+          "Output validation failed for [operator: {}, plan node ID: {}]: {}",
+          op.operatorType(),
+          op.planNodeId(),
+          e.what());
+    }
+  }
+}
+
+StopReason Driver::blockDriver(
+    const std::shared_ptr<Driver>& self,
+    size_t blockedOperatorId,
+    ContinueFuture&& future,
+    std::shared_ptr<BlockingState>& blockingState,
+    CancelGuard& guard) {
+  auto* op = operators_[blockedOperatorId].get();
+  VELOX_CHECK(
+      future.valid(),
+      "The operator {} is blocked but blocking future is not "
+      "set by isBlocked method.",
+      op->operatorType());
+
+  blockedOperatorId_ = blockedOperatorId;
+  blockingState = std::make_shared<BlockingState>(
+      self, std::move(future), op, blockingReason_);
+  guard.notThrown();
+  return StopReason::kBlock;
```

Comment:
Since blockDriver always return StopReason::kBlock, shall we just change it to return void which only set the driver states? Thanks!
```
blockDriver();
return StopReason::kBlock;
```

Reply:
@xiaoxmeng, thanks for the suggestion. It's just because it always return StopReason::kBlock that I decided to put it into blockDriver, such that the callers only need to call it on one line. It makes the operations to mark a driver blocked a coherent block so the callers of it doesn't need to handle it separately and doesn't have to spend time thinking what state it should be and if it is correct or not. So I prefer to keep it this way, unless there's something more convincing that I didn't notice. Thanks!

Reply:
If we make it a function return, people might guess it could return different stop reasons from this function when first read this code. That's why I think explicit return StopReason::kBlock by user seems better.